### ISSUE TITLE
Base rule extension: [`eqeqeq`] type-aware `smart` option

### ISSUE THREAD

**Issue by Rudxain** [typescript-eslint/typescript-eslint#10527]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+base+rule+extension%22) and found none that match my proposal.
- [X] I have searched the [current extension rule list](https://typescript-eslint.io/rules/?=extension#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the base rule

https://eslint.org/docs/latest/rules/eqeqeq

### Description

If the `"smart"` option is set, the new rule should only allow `==` if it's statically guaranteed to have identical semantics to `===`

### Fail

```TypeScript
declare const x: number | string;
declare const y: number | string;

// TS allows this, because of the overlap
x == y
```

### Pass

```TypeScript
// ESLint assumes both are `unknown`
Math.random() == Math.random()
```

**Comment by Josh-Cena** [comment]
Why would you ever want to use `==` outside of `== null`?

**Comment by bradzacher** [comment]
I agree with Josh as other maintenance team do as well from the reactions

- This is not a common pattern at all -- the vast, vast, vast majority codebases intentionally never use `==` except for the null case.
- I don't think there's any real broader want for people to use one less character.
    - I.e. this very much fails the "... my proposal would be useful to the broader TypeScript community."
- This is would be a stylistic change and thus is very opinionated.
- This would be injecting type info into a base rule and thus be a pretty expensive-to-maintain extension rule

For the above reasons we shall reject this proposal.

**Comment by Rudxain** [comment]
`BigInt`s. What about an exception for comparing them with `Number`s?

**Comment by kirkwaiblinger** [comment]
> > Why would you ever want to use `==` [...]?
> 
> `BigInt`s. What about an exception for comparing them with `Number`s?

Fun trivia, but this doesn't move the needle at all on Brad's assessment. Usage of bigints is exceedingly rare. And I'm guessing you don't have a real-world use case for `number == bigint` either or else you'd probably know [it's considered a TS error](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=GYVwdgxgLglg9mABBOBbADgQwE4FMBCmAJgBRgBciYIqARrtgDSK0yWsDmMYUAlO3DgAbXJiQBvAFCJEeKCGxIkAXmUsYAbkkBfSZNCRYCZGix4A4oNIUqNek3XsYXHvxaCRYxFJlyFSADk7BgA6GABnAEkeXA4GMl5EADIkxHxnaKgExFU1Vi1tIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false) 😁. (FWIW, I didn't know this until just now).

I'd assume this is the most robust way to compare numbers and bigints, but in truth I don't actually know.
```ts
function compare(n: number, bi: bigint): boolean {
  return Number.isInteger(n) && n === bi;
}
```

([playground](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=GYVwdgxgLglg9mABBOBbADgQwE4FMBCmAJgBRgBciYIqARrtgDSK0yWsDmMYUAlO3DgAbXJiQBvAFCJEeKCGxIkAXmUsYAbkkBfSZNCRYCZGix4A4oNIUqNek3XsYXHvxaCRYxFJlyFSADk7BgA6GABnAEkeXA4GMl5EADIkxHxnaKgExFU1Vi1tIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false))

**Comment by kirkwaiblinger** [comment]
(see also https://github.com/microsoft/TypeScript/issues/30540)

**Comment by Rudxain** [comment]
> Usage of bigints is exceedingly rare

That's unfortunate :( . In languages like Python, they are the norm.

> it's considered a TS error

Unless it's a union


```ts
// this guarantees no extra memory will be allocated,
// as `BigInt(Number.MAX_VALUE)` is 1025 bits
const cmp = (a: number | bigint, b: number | bigint) => a == b
```

**Comment by kirkwaiblinger** [comment]
> > Usage of bigints is exceedingly rare
> 
> That's unfortunate :( . In languages like Python, they are the norm.

Agreed, but sadly, and unlike Python, the language really doesn't make it easy to use `bigint`s though, since it doesn't accept bigints in the place of numbers. For example
 ```js
[0n, 1n, 2n].sort((a, b) => a - b); // Uncaught TypeError: Cannot convert a BigInt value to a number
[0n, 1n, 2n].sort((a, b) => Math.sign(a - b)); // Uncaught TypeError: Cannot convert a BigInt value to a number
```
Or, `a / b` does float division for two floats, integer division for two bigints, and an exception for a mix. Unlike python(3) where `a / b` is always a float and `a // b` is integer division.

The design in JS is that numbers and bigints are largely immiscible except by explicit conversion 🤷‍♂️ .

As far as this issue, the analysis is still basically
- very few codebases use a single bigint
- those that do are by no means required to use `==`, and it's an upstream TS error to _intentionally_ use `==` for its `bigint`-to-`number` behavior. 

Both of these more or less independently make any point about bigints DOA as far as weighing the proposal for a type-aware eqeqeq variant in typescript-eslint.

***

FWIW, it's quite an easy rule to write yourself if you have a usecase for this, though! In fact we've pretty much done this ourselves with our internal lint rule `eqeq-nullish`

https://github.com/typescript-eslint/typescript-eslint/blob/0d65f179ac88ab6bb60cd3fcb1ad67bdc725a9d1/packages/eslint-plugin-internal/src/rules/eqeq-nullish.ts#L1-L93
